syntax/printer.go

```diff
@@ -39,29 +49,39 @@ func SpaceRedirects(p *Printer) { p.spaceRedirects = true }
 // Note that this feature is best-effort and will only keep the
 // alignment stable, so it may need some human help the first time it is
 // run.
-func KeepPadding(p *Printer) {
-	p.keepPadding = true
-	p.cols.Writer = p.bufWriter.(*bufio.Writer)
-	p.bufWriter = &p.cols
+func KeepPadding() PrinterOption {
+	return func(p *Printer) {
+		p.keepPadding = true
+		p.cols.Writer = p.bufWriter.(*bufio.Writer)
+		p.bufWriter = &p.cols
+	}
 }
 
 // Minify will print programs in a way to save the most bytes possible.
 // For example, indentation and comments are skipped, and extra
 // whitespace is avoided when possible.
-func Minify(p *Printer) { p.minify = true }
+func Minify() PrinterOption {
+	return func(p *Printer) { p.minify = true }
+}
 
 // NewPrinter allocates a new Printer and applies any number of options.
-func NewPrinter(options ...func(*Printer)) *Printer {
+func NewPrinter(opts ...PrinterOption) *Printer {
 	p := &Printer{
 		bufWriter: bufio.NewWriter(nil),
 		tabWriter: new(tabwriter.Writer),
 	}
-	for _, opt := range options {
+	for _, opt := range opts {
 		opt(p)
 	}
 	return p
 }
 
+func (p *Printer) SetOption(opts ...PrinterOption) {
```

Comment:
not a big fan of `SetOption`; the option type is already pretty trivial to apply. why the extra API?

Reply:
Implicit docs and clear api for users.
Without this you would need to add explicit example on how to apply option to existing printer. And also answer your newly hired juniors why syntax for applying opts to new printer and existing differ.

```
[kaey@kpc sh]$ go doc ./syntax printer
package syntax // import "mvdan.cc/sh/v3/syntax"

type Printer struct {
        // Has unexported fields.
}
    Printer holds the internal state of the printing mechanism of a program.

func NewPrinter(opts ...PrinterOption) *Printer
func (p *Printer) Print(w io.Writer, node Node) error
func (p *Printer) SetOption(opts ...PrinterOption)


[kaey@kpc sh]$ go doc ./syntax printeroption
package syntax // import "mvdan.cc/sh/v3/syntax"

type PrinterOption func(*Printer)
    PrinterOption is a function which can be passed to NewPrinter or
    Printer.SetOption() to alter its behaviour.

func BinaryNextLine() PrinterOption
func Indent(spaces uint) PrinterOption
func KeepPadding() PrinterOption
func Minify() PrinterOption
func SpaceRedirects() PrinterOption
func SwitchCaseIndent() PrinterOption
```

Reply:
Gotcha. I think an example would be fine. With the addition of SetOption, there are now *three* ways to set up options for a new printer, and that's definitely bad design. One alternative would be to not accept any options in New, but in my opinion that's worse.

> And also answer your newly hired juniors why syntax for applying opts to new printer and existing differ.

That's just how it is if we want to allow New to take options. Juniors would most likely not be applying options to existing printers, so I dont' think this point is important. As said before, we can add an example if needed.

Reply:
Third way is really hard to discover and there is no need for users to use it.
Also there is another design where it's not possible at all - move config into separate unexported struct and define option like this `type PrinterOption func(*printerConfig)`. But I don't see the need to do this here.

>One alternative would be to not accept any options in New

This would make the most common usage more clunky, I don't see any need to do this.

Reply:
> But I don't see the need to do this here.

> This would make the most common usage more clunky,

We agree on those two points, at least :)

I still don't think SetOption is necessary. I see the point about discoverability on godoc, but we can accomplish that with a bit of docs and an example too. An extra API, particularly when it doesn't enable any new features, seems like a bad option to me. Please just revert; that change isn't necessary to use named type options.

Reply:
Done